File: ocaml/lib/day2.ml
open Base
open Stdex

type input = (char * char) list 

type shape = Rock | Paper | Scissors [@@deriving equal, show]
type result = Win | Lose | Draw [@@deriving equal, show]

type rule = { 
  them : shape; 
  us : shape; 
  result : result 
} [@@deriving show]

let rules = [ 
  { them = Rock; us = Rock; result = Draw };
  { them = Rock; us = Paper; result = Win };
  { them = Rock; us = Scissors; result = Lose };
  { them = Paper; us = Rock; result = Lose };
  { them = Paper; us = Paper; result = Draw };
  { them = Paper; us = Scissors; result = Win };
  { them = Scissors; us = Rock; result = Win };
  { them = Scissors; us = Paper; result = Lose };
  { them = Scissors; us = Scissors; result = Draw };
]

let char_to_shape = function
  | 'A' | 'X' -> Rock
  | 'B' | 'Y' -> Paper
  | 'C' | 'Z' -> Scissors
  | _ -> failwith "Invalid shape"

let char_to_result = function
  | 'X' -> Lose
  | 'Y' -> Draw
  | 'Z' -> Win
  | _ -> failwith "Invalid result"

let result_to_score = function
  | Lose -> 0
  | Draw -> 3
  | Win -> 6

let shape_to_score = function
  | Rock -> 1
  | Paper -> 2
  | Scissors -> 3

let parse_line line = 
  let (a, b) = String.lsplit2_exn ~on:' ' line in (a.[0], b.[0])

let parse_input lines = List.map ~f:parse_line lines

let find_result (a, b) =
  let them = char_to_shape a in
  let us = char_to_shape b in
  let r = List.find_exn rules ~f:(fun r -> equal_shape r.us us && 
                                           equal_shape r.them them)
  in r.result

let find_shape (a, b) =
  let them = char_to_shape a in
  let result = char_to_result b in
  let r = List.find_exn rules ~f:(fun r -> equal_shape r.them them && 
                                           equal_result r.result result)
  in r.us

let score1 turn =
  let r = find_result turn in
  let s = char_to_shape (snd turn) in
  (result_to_score r) + (shape_to_score s)

let score2 turn =
  let s = find_shape turn in
  let r = char_to_result (snd turn) in
  (result_to_score r) + (shape_to_score s)

let part1 turns = List.map ~f:score1 turns |> List.sum
let part2 turns = List.map ~f:score2 turns |> List.sum

File: haskell/day2.hs
module Day2 where

import Data.List
import Util
import qualified Aoc

data Shape = Rock | Paper | Scissors deriving Eq
data Result = Win | Lose | Draw deriving Eq
type Rule = (Shape, Shape, Result) 

rules = [
  (Rock, Rock, Draw),
  (Rock, Paper, Win),
  (Rock, Scissors, Lose),
  (Paper, Rock, Lose),
  (Paper, Paper, Draw),
  (Paper, Scissors, Win),
  (Scissors, Rock, Win),
  (Scissors, Paper, Lose),
  (Scissors, Scissors, Draw)]

parseShape :: Char -> Shape
parseShape 'A' = Rock
parseShape 'B' = Paper
parseShape 'C' = Scissors
parseShape 'X' = Rock
parseShape 'Y' = Paper
parseShape 'Z' = Scissors

parseResult :: Char -> Result
parseResult 'X' = Lose
parseResult 'Y' = Draw
parseResult 'Z' = Win

resultToScore :: Result -> Int
resultToScore Lose = 0
resultToScore Draw = 3
resultToScore Win = 6

shapeToScore :: Shape -> Int
shapeToScore Rock = 1
shapeToScore Paper = 2
shapeToScore Scissors = 3

parse :: [String] -> [(Char, Char)]
parse = map (\s -> (head s, last s))

findResult (a, b) =
  let theirs = parseShape a in
  let ours = parseShape b in
  head [r | (s1, s2, r) <- rules, s1 == theirs && s2 == ours]

findShape (a, b) =
  let theirs = parseShape a in
  let result = parseResult b in
  head [s2 | (s1, s2, r) <- rules, s1 == theirs && r == result]

part1 = sum . (map score)
  where score (a,b) =
          let r = findResult (a,b) in
          let s = parseShape b in
          (resultToScore r) + (shapeToScore s)
          
part2 = sum . (map score)
  where score (a,b) =
          let s = findShape (a,b) in
          let r = parseResult b in
          (resultToScore r) + (shapeToScore s)

run = Aoc.run 2 parse part1 part2

